src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.BatchingDocumentCollection.cs

```diff
@@ -537,40 +543,63 @@ public void ReorderFiles(ImmutableArray<string> filePaths)
+                Func<Solution, ImmutableArray<DocumentInfo>, Solution> addDocuments,
+                WorkspaceChangeKind addDocumentChangeKind,
+                Func<Solution, ImmutableArray<DocumentId>, Solution> removeDocuments,
+                WorkspaceChangeKind removeDocumentChangeKind,
+                ProjectId projectId,
+                ImmutableArray<DocumentInfo> documentsAddedInBatch,
+                ImmutableArray<DocumentId> documentsRemovedInBatch,
+                ImmutableList<DocumentId>? orderedDocumentsInBatch,
+                Func<DocumentId, SourceTextContainer?> getContainer)
             {
-                Contract.ThrowIfNull(documentInfo.FilePath, "We shouldn't be adding documents without file paths.");
-                documentFileNamesAdded.Add(documentInfo.FilePath);
+                // Document adding...
+                solutionChanges.UpdateSolutionForDocumentAction(
+                    newSolution: addDocuments(solutionChanges.Solution, documentsAddedInBatch),
+                    changeKind: addDocumentChangeKind,
+                    documentIds: documentsAddedInBatch.Select(d => d.Id));
 
-                if (_sourceTextContainersToDocumentIds.TryGetKey(documentInfo.Id, out var textContainer))
+                foreach (var documentInfo in documentsAddedInBatch)
                 {
-                    documentsToOpen.Add((documentInfo.Id, textContainer));
+                    Contract.ThrowIfNull(documentInfo.FilePath, "We shouldn't be adding documents without file paths.");
+                    documentFileNamesAdded.Add(documentInfo.FilePath);
+
+                    var textContainer = getContainer(documentInfo.Id);
+                    if (textContainer != null)
+                    {
+                        documentsToOpen.Add((documentInfo.Id, textContainer));
+                    }
                 }
-            }
 
-            ClearAndZeroCapacity(_documentsAddedInBatch);
+                // Document removing...
+                solutionChanges.UpdateSolutionForRemovedDocumentAction(removeDocuments(solutionChanges.Solution, [.. documentsRemovedInBatch]),
+                    removeDocumentChangeKind,
+                    documentsRemovedInBatch);
 
-            // Document removing...
-            solutionChanges.UpdateSolutionForRemovedDocumentAction(removeDocuments(solutionChanges.Solution, [.. _documentsRemovedInBatch]),
-                removeDocumentChangeKind,
-                _documentsRemovedInBatch);
+                // Update project's order of documents.
+                if (orderedDocumentsInBatch != null)
+                {
+                    solutionChanges.UpdateSolutionForProjectAction(
+                        projectId,
+                        solutionChanges.Solution.WithProjectDocumentsOrder(projectId, orderedDocumentsInBatch));
+                }
+            }
+        }
 
+        internal void ClearBatchState()
```

Comment:
needs to happen outside the sln update.

Reply:
doc this.  also doc the caller who eventually calls this, and have it explain that we explicitly only clear once we've successfully pushed the sln update.